### ISSUE TITLE
Creating a pinned GCHandle to an array of a blittable type may fail

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/runtime#48907]
### Description

Just for the fun of it, try figuring out under which conditions this may fail (answer further below):

```csharp
static void PinArray (int[] array)
{
	// This may throw "System.ArgumentException: Object contains non-primitive or non-blittable data. (Parameter 'value')"
	var handle = GCHandle.Alloc (array, GCHandleType.Pinned);
	handle.Free ();
}
```

Surely an `int[]` is a blittable type?

Let's try:

```csharp
PinArray (new int [] { 1, 2, 3 });
```

Yes, that works just fine. However, the array might actually be an array of enum values, which isn't blittable (even though it's identical enough to int[] for the runtime to accept casting from one to the other):

```csharp
// This does not work...
// "System.ArgumentException: Object contains non-primitive or non-blittable data. (Parameter 'value')"
PinArray ((int[]) (object) new E [] { E.A, E.B, E.C });

enum E { A, B, C, }
```

This seems a bit broken... when the input is `int[]`, and the runtime allows you to cast an array of enums to `int[]` because the memory layout is identical, then it shouldn't fail sometimes because of some fine print somewhere that says that enums aren't blittable.

<details>
<summary>Full test case</summary>

```csharp
using System;
using System.Runtime.InteropServices;

class App {
	public static void Main	 ()
	{
		// This works just fine:
		PinArray (new int [] { 1, 2, 3 });
		// This does not...
		PinArray ((int[]) (object) new E [] { E.A, E.B, E.C });
	}

	static void PinArray (int[] array)
	{
		// This may throw "System.ArgumentException: Object contains non-primitive or non-blittable data. (Parameter 'value')"
		var handle = GCHandle.Alloc (array, GCHandleType.Pinned);
		handle.Free ();
	}

	enum E {
		A,
		B,
		C,
	}
}
```

</details >

### Configuration

```
 $ dotnet --version
6.0.100-preview.2.21114.3
```

### Regression?

No, as far as I know it's been this way since forever.

**Comment by tannergooding** [comment]
This might be related to the `DISPARAM` scenario that @AaronRobinsonMSFT mentioned here: https://github.com/dotnet/runtime/issues/12258#issuecomment-472991760.

There are a few spots where the marshalling expectations for types holds true for regular `P/Invoke`, but not for COM Interop.

**Comment by john-h-k** [comment]
Looks like this code is causing the issue

```cpp
BASEARRAYREF asArray = (BASEARRAYREF)ObjectToOBJECTREF(obj);
if (CorTypeInfo::IsPrimitiveType(asArray->GetArrayElementType()))
    FC_RETURN_BOOL(TRUE);

TypeHandle th = asArray->GetArrayElementTypeHandle();
if (!th.IsTypeDesc())
{
    MethodTable *pMT = th.AsMethodTable();
    if (pMT->IsValueType() && pMT->IsBlittable())
        FC_RETURN_BOOL(TRUE);
}

FC_RETURN_BOOL(FALSE);
```

For some reason it seems enums are not considered blittable.


If one of the unused method table flags was changed to be `enum_flag_AutoLayout`, then this method could be trivially rewritten (either in the runtime or even in managed code and avoid the transition) to:

```cs
bool IsPinnable(object o)
{
    var pMT = GetMethodTable(o);
    var isPinnable = !pMT->ContainsGCPointers && !pMT->HasAutoLayout;
    GC.KeepAlive(o);
    return isPinnable;
}```

**Comment by AaronRobinsonMSFT** [comment]
>For some reason it seems enums are not considered blittable.

That is true. Enums are not technically blittable since they can be backed by non-blittable types (e.g., `bool` and `char`) - see https://github.com/dotnet/runtime/issues/12258#issuecomment-473048186.

**Comment by AaronRobinsonMSFT** [comment]
This isn't to say we can't consider this issue but in practice I would likely consider this similar in priority to https://github.com/dotnet/runtime/issues/12258.

**Comment by MichalStrehovsky** [comment]
> That is true. Enums are not technically blittable since they can be backed by non-blittable types (e.g., `bool` and `char`) - see https://github.com/dotnet/runtime/issues/12258#issuecomment-473048186.

`bool` and `char` don't have the problem with being castable by size to a pinnable array type (so if I have a `MyEnum[]` where `MyEnum` is an enum based on `bool`, I will not be able to cast it to `byte[]` and get an "unexpected" ArgumentException when trying to pin the array).

Only int8/16/32/64/nint and uint8/16/32/64/nuint are castable by size and can hit the issue OP is discussing.

**Comment by jkotas** [comment]
Fixed by #68694
